### ISSUE TITLE
KeyError in Spack Environments

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#13553]
I have an environment already installed. I want to update that environment by running `spack concretize --force`, but I'm seeing a KeyError when I try to.

### Steps to reproduce the issue

```console
$ spack concretize
$ spack install
# update one of the packages in the environment
$ spack concretize --force
```

### Error Message

```console
$ spack -d concretize -f
Traceback (most recent call last):
  File "/Users/Adam/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 653, in main
    ev.activate(env, args.use_env_repo)
  File "/Users/Adam/spack/lib/spack/spack/environment.py", line 163, in activate
    cmds += env.add_default_view_to_shell(shell)
  File "/Users/Adam/spack/lib/spack/spack/environment.py", line 1090, in add_default_view_to_shell
    for _, spec in self.concretized_specs():
  File "/Users/Adam/spack/lib/spack/spack/environment.py", line 1215, in concretized_specs
    yield (s, self.specs_by_hash[h])
KeyError: 'pxw6s2yxrbws3sg5jxzosiqc3xiisane'
```

### Information on your system

I've been able to reproduce this on both macOS 10.15.1 and Cray CNL5. The only workaround I've found so far is to `rm spack.lock` and start from scratch in the environment.

@alalazo This may be related to past problems we've encountered where the `package.py` used to install a package and generate module files is different from the current `package.py`. In this case, the only major change I know of is that `sqlite` had a new version added, so that may be the problem.

**Comment by phanousk** [comment]
I have the same issue with clean spack installation:

>     git clone https://github.com/spack/spack spack-0.13.1
>     cd spack-0.13.1
>     git checkout releases/v0.13
>     spack install zlib@1.2.8
>     spack env create userenv-0131
>     spack env activate userenv-0131
>     spack add zlib@1.2.8
>     spack add libxml2
>     spack install

```Traceback (most recent call last):
  File "/cvmfs/software.myrepo/tests/software_basics/spack-0.13.1/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/cvmfs/software.myrepo/tests/software_basics/spack-0.13.1/lib/spack/spack/main.py", line 653, in main
    ev.activate(env, args.use_env_repo)
  File "/cvmfs/software.myrepo/tests/software_basics/spack-0.13.1/lib/spack/spack/environment.py", line 162, in activate
    cmds += env.add_default_view_to_shell(shell)
  File "/cvmfs/software.myrepo/tests/software_basics/spack-0.13.1/lib/spack/spack/environment.py", line 1119, in add_default_view_to_shell
    for _, spec in self.concretized_specs():
  File "/cvmfs/software.myrepo/tests/software_basics/spack-0.13.1/lib/spack/spack/environment.py", line 1244, in concretized_specs
    yield (s, self.specs_by_hash[h])
KeyError: '3yl7xdhjft4obodsxdavnem42q57q624'
```

**Comment by scheibelp** [comment]
@phanousk could you report this in a separate issue? I don't think your error is related to this. Note that if you are invoking `spack` without a full path, I would expect that you have to invoke `setup-env.sh` after cloning the Spack repo.

**Comment by phanousk** [comment]
> @phanousk could you report this in a separate issue? I don't think your error is related to this. Note that if you are invoking `spack` without a full path, I would expect that you have to invoke `setup-env.sh` after cloning the Spack repo.

Hi, I've tried today the same test with prior 'git pull' and can not simulate the problem again. So perhaps it's not necessary to file another issue.
Petr

**Comment by tgamblin** [comment]
@asmaahassan90 saw something like this with a slightly different sequence of operations.  I think it's the same issue and it's fixed by #14390.
